stub/src/main/java/io/grpc/stub/ClientCalls.java

```diff
@@ -122,81 +117,76 @@
+   * @return a future for the single response message.
    */
-  public static <ReqT, RespT> void asyncServerStreamingCall(
+  public static <ReqT, RespT> ListenableFuture<RespT> futureUnaryCall(
       ClientCall<ReqT, RespT> call,
-      ReqT param,
-      StreamObserver<RespT> responseObserver) {
-    asyncServerStreamingCall(call, param,
-        new StreamObserverToCallListenerAdapter<RespT>(call, responseObserver));
+      ReqT param) {
+    GrpcFuture<RespT> responseFuture = new GrpcFuture<RespT>(call);
+    asyncUnaryRequestCall(call, param, new UnaryStreamToFuture<RespT>(responseFuture));
+    return responseFuture;
   }
 
-  private static <ReqT, RespT> void asyncServerStreamingCall(
-      ClientCall<ReqT, RespT> call,
-      ReqT param,
-      ClientCall.Listener<RespT> responseListener) {
-    call.start(responseListener, new Metadata.Headers());
-    call.request(1);
+  /**
+   * Returns the result of calling {@link Future#get()} interruptably on a task known not to throw a
+   * checked exception.
+   *
+   * <p>If interrupted, the interrupt is restored before throwing a {@code RuntimeException}.
+   *
+   * @throws RuntimeException if {@code get} is interrupted
+   * @throws CancellationException if {@code get} throws a {@code CancellationException}
+   * @throws StatusRuntimeException if {@code get} throws an {@code ExecutionException}
+   */
+  private static <V> V getUnchecked(Future<V> future) {
     try {
-      call.sendPayload(param);
-      call.halfClose();
-    } catch (Throwable t) {
-      call.cancel();
-      throw Throwables.propagate(t);
+      return future.get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException e) {
+      throw Status.fromThrowable(e).asRuntimeException();
     }
   }
 
-  /**
-   * Executes a client-streaming call with a blocking {@link Iterator} of request messages.
-   * @return the single response value.
-   */
-  public static <ReqT, RespT> RespT blockingClientStreamingCall(
+  private static <ReqT, RespT> void asyncUnaryRequestCall(
+      ClientCall<ReqT, RespT> call, ReqT param, StreamObserver<RespT> responseObserver,
+      boolean streamingResponse) {
+    asyncUnaryRequestCall(call, param,
+        new StreamObserverToCallListenerAdapter<RespT>(call, responseObserver, streamingResponse));
+  }
+
+  private static <ReqT, RespT> void asyncUnaryRequestCall(
       ClientCall<ReqT, RespT> call,
-      Iterator<ReqT> clientStream) {
-    GrpcFuture<RespT> responseFuture = new GrpcFuture<RespT>(call);
-    call.start(new UnaryStreamToFuture<RespT>(responseFuture), new Metadata.Headers());
+      ReqT param,
+      ClientCall.Listener<RespT> responseListener) {
+    call.start(responseListener, new Metadata.Headers());
+    // Initially ask for two responses from flow-control so that if a misbehaving server sends more
+    // than one responses, we can catch it.
+    call.request(2);
```

Comment:
The problem with not having the sanity check is that we will hang the stream. That doesn't really seem acceptable. I'd much rather not do the optimization if we can't easily keep the checks.